grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -228,6 +226,88 @@ private static void addUnvisitedNodeFragments(List<Fragment> plan,
+                }
+            }
+            nodesWithFixedCost.put(start, logInstanceCount);
+            start.setFixedFragmentCost(fragment.fragmentCost());
+
+        } else if (fragment.dependencies().isEmpty()) {
+            //fragments that should be done when a node has been visited
+            start.getFragmentsWithoutDependency().add(fragment);
+        }
+    }
+
+    private static void processFragmentWithDependencies(Map<NodeId, Node> allNodes, Fragment fragment) {
+        // it's either neq or value fragment
+        Node start = Node.addIfAbsent(NodeId.NodeType.VAR, fragment.start(), allNodes);
+        Node other = Node.addIfAbsent(NodeId.NodeType.VAR, fragment.dependencies().iterator().next(),
+                allNodes);
+
+        start.getFragmentsWithDependency().add(fragment);
+        other.getDependants().add(fragment);
+
+        // check whether it's value fragment
+        if (fragment.varProperty() instanceof ValueProperty) {
+            // as value fragment is not symmetric, we need to add it again
+            other.getFragmentsWithDependency().add(fragment);
+            start.getDependants().add(fragment);
+        }
+    }
+
+    private static void processSubFragment(List<Fragment> plan,
+                                           Map<NodeId, Node> allNodes,
+                                           Set<Node> connectedNodes,
+                                           Map<Node, Double> nodesWithFixedCost,
+                                           Set<Fragment> allFragments) {
+
+        Set<Fragment> validSubFragments = allFragments.stream().filter(fragment -> {
+            if (fragment instanceof InSubFragment) {
+                Node superType = Node.addIfAbsent(NodeId.NodeType.VAR, fragment.start(), allNodes);
+                if (nodesWithFixedCost.containsKey(superType) && nodesWithFixedCost.get(superType) > 0D) {
+                    return true;
+                }
+            }
+            return false;
+        }).collect(Collectors.toSet());
+
+        if (!validSubFragments.isEmpty()) {
+            final Set<Fragment> subFragmentSet = new HashSet<>();
+            validSubFragments.forEach(fragment -> {
+                Node superType = Node.addIfAbsent(NodeId.NodeType.VAR, fragment.start(), allNodes);
+                plan.add(fragment);
+                connectedNodes.add(superType);
+                nodesWithFixedCost.put(
+                        Node.addIfAbsent(NodeId.NodeType.VAR, fragment.end(), allNodes),
+                        nodesWithFixedCost.get(superType));
+                subFragmentSet.add(fragment);
+            });
+            subFragmentSet.forEach(inSubFragment -> {
+                allFragments.remove(inSubFragment);
+                allFragments.remove(Fragments.outSub(
+                        inSubFragment.varProperty(), inSubFragment.end(), inSubFragment.start()));
```

Comment:
You can replace this with `inSubFragment.getInverse()`

Reply:
Let me check if I have implemented this method for sub :-)